### DISCUSSION TITLE
Prometheus, and timing requests

### DISCUSSION THREAD

**Discussion by JensErikChristiansen** [nuxt/nuxt#17094] [Questions] [answered]
Hi!

I'm trying to use Prometheus for monitoring, and want to log the time taken for each request, as well as the status code. It looks like server/middleware only happens _before_ each request, but I also need to inspect the results _after_ the request has finished in order to stop the timer and get the status code. So the order would be:

1. start the timer
2. make the actual request
3. stop the timer and get status code
4. pass this data on to Prometheus

Any way we could do that without putting this code into each and every api? Thanks!

**Comment by Flanker72** [comment, marked as answer]
I use such code to handle the same scenario on the API requests:

/server/middleware/profiler.ts:
```ts
/**
 * Profile every API request.
 */
import logger from "~~/utils/logger";

export default defineEventHandler(async (event) => {
  const profiler = logger.startTimer();

  event.res.on("finish", async () => {
    profiler.done({ message: `${event.req.method} ${event.req.url}` });
  });
});

```

**Reply by JensErikChristiansen** [reply]
This is EXACTLY what I was looking for, thank you! Where did you find the documentation on events for `res`?

**Reply by pi0** [reply]
This is a nice solution! res events are documentes in Node.js: https://nodejs.org/api/http.html#event-finish

(BTW nitro also has an experimental timing flag. We might expose it later for cross platform timing hooks)

**Reply by JensErikChristiansen** [reply]
ah, so Nuxt is using h3's `defineEventHandler`, which in turn is passing in standard node http events?
